Docs/doc 95/doc 106/technical term metric (#4116)
great-expectations/great_expectations: docs/terms/metric.md

hoverText: A computed attribute of data such as the mean of a column.
---
import UniversalMap from '/docs/images/universal_map/_universal_map.mdx';
import TechnicalTag from '../term_tags/_tag.mdx';
import CreateHeader from '/docs/images/universal_map/_um_create_header.mdx';
import ValidateHeader from '/docs/images/universal_map/_um_validate_header.mdx';

<UniversalMap setup='inactive' connect='inactive' create='active' validate='active'/> 

## Overview

### Definition

> A Metric is a computed attribute of data such as the mean of a column.

Reviewer: I know we've talked about the notion that a Metric is just the answer to a question (what is the mean of my column, how many rows in my column are within the range 0<100), and I wonder whether we would want to mention that here at this stage. I believe Abe has worked on some content around this, and it felt like it could be major part of the narrative for GE.

Author: I've added that to the following section, since one of the Standards for theses docs is for the Definition section to mirror the Definition that shows up in hover text when a Technical Term is tagged in another document.

### Features and promises

Metrics are values derived from one or more <TechnicalTag relative="../" tag="batch" text="Batches" /> that can be used to evaluate <TechnicalTag relative="../" tag="expectation" text="Expectations" /> or to summarize the result of <TechnicalTag relative="../" tag="validation" text="Validation" />. A Metric could be a statistic, such as the minimum value of the column, or a more complex object, such as a histogram. Metrics are the core tool used to Validate data. 

### Relationship to other objects

> Metrics are generated and stored as part of running Expectations against a Batch (and can be referenced as such). For example, if you have an Expectation that the mean of a column fall within a certain range, the mean of the column must first be computed to see if its value is as expected.  The generation of Metrics involves <TechnicalTag relative="../" tag="execution_engine" text="Execution Engine" /> specific logic.  Metrics from previously run Expectation Suites can also be referenced through the use of <TechnicalTag relative="../" tag="evaluation_parameter" text="Evaluation Parameters" />.

Reviewer: It might be ambiguous/confusing to say that Metrics are stored. They are stored in memory so they can be used in the current validation. They can also be stored within ExpectationValidationResults, though not all computed metrics will be, and this will also depend on `result_format`.

They _can_ be stored in a Metric Store, though this functionality is still pretty experimental and not totally recommend (though design work for an upgrade is in-progress).

```suggestion
Metrics are generated and stored as part of running Expectations against a Batch (and can be referenced as such). For example, if you have an Expectation that the mean of a column falls within a certain range, the mean of the column must first be computed to see if its value is as expected.  The generation of Metrics involves <TechnicalTag relative="../" tag="execution_engine" text="Execution Engine" /> specific logic.  Metrics from previously run Expectation Suites can also be referenced through the use of <TechnicalTag relative="../" tag="evaluation_parameter" text="Evaluation Parameters" />.
```

Author: I made this edit and clarified how some metrics, based on `result_format` of Validation Results, can be accessed through in memory Validation Results by Actions, which can then store them so that they can be referenced from the Validation Results Store later.

## Use cases

<CreateHeader/>

Metrics are generated in accordance with the requirements of an Expectation when an Expectation is evaluated.  This includes Expectations that are evaluated as part of the [interactive process for creating Expectations](../guides/expectations/how_to_create_and_edit_expectations_with_instant_feedback_from_a_sample_batch_of_data.md) and the [Profiler process for creating Expectations](../guides/expectations/how_to_create_and_edit_expectations_with_a_profiler.md).  

Past Metrics can also be accessed by some Expectations through Evaluation Parameters.  However, when you are creating Expectations there may not be past Metrics to provide.  In these cases, it is possible to define a temporary value that the Evaluation Parameter can use in place of the missing past Metric.

<ValidateHeader/>

> <TechnicalTag relative="../" tag="checkpoint" text="Checkpoints" /> Validate data by running the Expectations in one or more <TechnicalTag relative="../" tag="expectation_suite" text="Expectation Suite" />.  In the process, Metrics will be generated.  These Metrics will be passed to the Actions in the Checkpoint's `action_list` as part of the <TechnicalTag relative="../" tag="validation_result" text="Validation Results" /> for the Expectations, and will be stored in a <TechnicalTag relative="../" tag="metric_store" text="Metric Store" /> if the `StoreMetricsAction` is one of the <TechnicalTag relative="../" tag="validation_action" text="Actions" /> in the Checkpoint's `action_list`.

Reviewer: I'm not sure it's accurate to say that Metrics are passed to Action in the `action_list`. I think it's more that ValidationResults can be saved using an Action, and these ValidationResults will include some Metrics.

Author: I added this clarification.

## Features

> ### Metrics are the core tool used to Validate data

Reviewer: It feels a bit odd to refer to Metrics as a tool. Suggestion below, but feel free to modify to something else.

```suggestion
### Metrics are core to the Validation of data
```

When an Expectation should be evaluated, Great Expectations collects all the Metrics requested by the Expectation and provides them to the Expectation's validation logic. Most validation is done by comparing values from a column or columns to a Metric associated with the Expectation being evaluated.

### Past Metrics are available to other Expectations and Data Docs

> An Expectation can also expose Metrics, such as the observed value of a useful statistic via an Expectation Validation Result, where <TechnicalTag relative="../" tag="data_docs" text="Data Docs" /> -- or other Expectations -- can use them.  This is done through an Action (to which the Expectation's Validation Result has been passed) which will save them to a Metric Store.  The Action in question is the `StoreMetricsAction`.  You can view the implementation of this Action in the `great_expectations/checkpoint/actions.py` file of a Great Expectations deployment, or the [same file in our GitHub.](https://github.com/great-expectations/great_expectations/blob/0312642755f6003c70623e9aa3ceed1020373dac/great_expectations/checkpoint/actions.py#L905)

Reviewer: Would it be fair to just include to the link to GitHub, rather than referencing the file path? Including both seems unnecessary.

```suggestion
An Expectation can also expose Metrics, such as the observed value of a useful statistic via an Expectation Validation Result, where <TechnicalTag relative="../" tag="data_docs" text="Data Docs" /> -- or other Expectations -- can use them.  This is done through an Action (to which the Expectation's Validation Result has been passed) which will save them to a Metric Store.  The Action in question is the `StoreMetricsAction`.  You can view the implementation of this Action in the `great_expectations/checkpoint/actions.py` file of a Great Expectations deployment, or the [same file in our GitHub.](https://github.com/great-expectations/great_expectations/blob/0312642755f6003c70623e9aa3ceed1020373dac/great_expectations/checkpoint/actions.py#L905)
```

Author: I included the file path for people who prefer looking at a local copy of the code they are using rather than the GitHub on the premise that the local copy will definitely contain what they are working with while the GitHub may have been updated and not accurately reflect whatever it is that they are looking into the file to discover.  I thought this might also be relevant to users who are using a legacy deployment and need to be able to compare the code that is currently operating for them with the code in GitHub to see if they are encountering an issue due to having an out of date version of the code.

Reviewer: Fair enough. I guess I assume that user who look at the link can deduce the filepath, but if you think this is helpful, that's fair.

## API basics

> ### How to access

Reviewer: In this section, you may also want to describe the flow of using `validator.get_metric/s`.

Validation Results can expose Metrics that are defined by specific Expectations that have been validated, called "Expectation Defined Metrics." To access those values, you address the Metric as a dot-delimited string that identifies the value, such as `expect_column_values_to_be_unique.success`or `expect_column_values_to_be_between.result.unexpected_percent`. These Metrics may be stored in a Metrics Store.

> It is important to note that these Metrics are *NOT* available for compute on raw data directly as other Metrics are.

Reviewer: What does this mean?

Author: I'm not entirely certain myself, tbh.  This section was pulled from the original Core Concepts document on Metrics.  When I first read it, I thought that it meant that dot-deliminated metrics couldn't be included in evaluation parameter expressions.

Reviewer: Ah. Maybe there should just be a note here that Metrics need to be calculated, and you can't just reference a Metric with calculating it first? Am I understanding correctly?

A `metric_kwargs_id` is a string representation of the Metric Kwargs that can be used as a database key. For simple cases, it could be easily readable, such as `column=Age`, but when there are multiple keys and values or complex values, it will most likely be a md5 hash of key/value pairs. It can also be `None` in the case that there are no kwargs required to identify the Metric.

The following examples demonstrate how Metrics are defined:

```python title="Python code"

> res = df.expect_column_values_to_be_in_set(

Reviewer: Minor nit, but could we do 
```suggestion
res = validator.expect_column_values_to_be_in_set(
```
We used to use `df` with V2, and I believe we should be using `validator` with `V3`.

    "color",
    ["red", "green"]
)
res.get_metric(
    "expect_column_values_to_be_in_set.result.missing_count",
    column="color"
)
```

> See the [How to configure a MetricsStore](../guides/setup/configuring_metadata_stores/how_to_configure_a_metricsstore.md) guide for more information.

Reviewer: It would be good to confirm, but we may not want to advertise the MetricsStore so heavily, and we may want a note that it is experimental, and an upgrade will be forthcoming. (Not just specific to this location, but wherever the existing MetricsStore is referenced.)

### How to create

> Metrics are produced using logic specific to the Execution Engine associated with the Datasource that provides the data for the Batch Request/s that the Metric is calculated for.  That logic that is defined in a `MetricProvider`. When a `MetricProvider` class is first encountered, Great Expectations will register the Metric and any methods that it defines as able to produce Metrics.  This means that after Validation, that Metric can then be referenced in Evaluation Parameters. Furthermore, the methods associated with a Metric can also be referenced through dot notation using the Metric as a base (as shown above in the [section on accessing Metrics](#how-to-access)). 

Reviewer: I don't think this is quite right. I don't think the reference to Evaluation Parameters should be here. Evaluation Parameters can be used to get an already calculated Metric, whereas registration is more for the method of Metric calculation. The registered metric will then be able to be used with `validator.get_metric()` / `validator.get_metrics()`.

Author: I see.  I misunderstood some of the documentation I was collating for this, then.  I've updated the section accordingly.

#### Types of MetricProvider Functions

The diagram below shows the relationship between different types of MetricProvider functions.

> ![Image](../images/metricprovider.png)

Reviewer: Currently, this might be confusing - I think to include this diagram, a lot more explanation is necessary. I think this might be more useful in docs on creating new Metrics, and might be a bit much for this page.

Author: Fair.  I've pulled this section and added a JIRA to ensure that the diagram is included in a document on creating new Metrics.

You can view the implementations of MetricProvider subclasses in the `great_expectations/expectations/metrics` folder of a Great Expectations deployment, or [in the same folder of our GitHub](https://github.com/great-expectations/great_expectations/tree/develop/great_expectations/expectations/metrics).

### Configuration

Configuration of Metrics is applied when they are defined as part of an Expectation.

#### Metrics naming conventions

Metrics can have any name. However, for the "core" Great Expectations Metrics, we use the following conventions:

> * For **aggregate Metrics**, such as the mean value of a column, we use describe the domain and name of the statistic, such as `column.mean` or `column.max`.

Reviewer: ```suggestion
* For **aggregate Metrics**, such as the mean value of a column, we use the domain and name of the statistic, such as `column.mean` or `column.max`.
```

* For **map Metrics**, which produce values for individual records or rows, we define the domain using the prefix "column_values" and use several consistent suffixes to provide related Metrics. For example, for the Metric that defines whether specific column values fall into an expected set, several related Metrics are defined:
    * `column_values.in_set.unexpected_count` provides the total number of unexpected values in the domain.
    * `column_values.in_set.unexpected_values` provides a sample of unexpected_values; "result_format" is one of its
      value_keys to determine how many values should be returned.
    * `column_values.in_set.unexpected_rows` provides full rows for which the value in the domain column was unexpected
    * `column_values.in_set.unexpected_value_counts` provides a count of how many times each unexpected value occurred

Additionally, to facilitate optimized computation of Metrics, we use **Metric Partials** which define partially-parameterized functions that are necessary to build a desired Metric.

* For aggregate Metrics, we often use an `ExecutionEngine` specific function with the suffix `.aggregate_fn`, such as `column.max.aggregate_fn`.

> * For map Metrics, to compute `column_values.in_set.unexpected_count`, we will rely on a **condition** called `column_values.in_set.condition`.

Reviewer: This could be pretty confusing at this point.

I think there are two different things here which it's important  to differentiate:
- Metric implementations - this is where registration comes in, and where knowing about `aggregate_fn` etc would come in. This feels somewhat advanced, and would mostly come into play when implementing a new Expectation/Metric.
- Metric values - These are the metrics that have been calculated, have been stored in Validation Results, and could be accessed using Evaluation Parameters.

We may want a bit of discussion about Metric implementations here, but there's a lot of detail and complexity which could be overwhelming if a user is coming to this page to understand how to get `unexpected_values`, for instance. I could see having an "Advanced" header, and placing this below that on this page, but then I think a lot more explanation will be necessary.

Author: I've pulled this section and added it to the JIRA ticket with things to include in a document about creating new metrics.
